src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -377,38 +379,285 @@ internal PhysicalLocationVersionOne CreatePhysicalLocation(Region v2Region)
+                        {
+                            // Embedded binary file content
+
+                            byte[] content = Convert.FromBase64String(fileData.Contents.Binary);
+                            stream = new MemoryStream(content);
+                        }
+                        else if (fileData.Contents.Text != null)
+                        {
+                            // Embedded text file content
+
+                            byte[] content = encoding.GetBytes(fileData.Contents.Text);
+                            stream = new MemoryStream(content);
+                        }
+                        else if (uri.IsAbsoluteUri && uri.Scheme == Uri.UriSchemeFile && File.Exists(uri.LocalPath))
+                        {
+                            // External source file
+
+                            try
+                            {
+                                stream = new FileStream(uri.LocalPath, FileMode.Open);
+                            }
+                            catch (FileNotFoundException ex)
+                            {
+                                failureReason = $"File '{uri.LocalPath}' could not be found: {ex.ToString()}";
+                            }
+                            catch (IOException ex)
+                            {
+                                failureReason = $"File '{uri.LocalPath}' could not be read: {ex.ToString()}";
+                            }
+                            catch (SecurityException ex)
+                            {
+                                failureReason = $"File '{uri.LocalPath}' could not be accessed: {ex.ToString()}";
+                            }
+                        }
+                    }
+                    else
+                    {
+                        failureReason = $"Encoding for file '{uri.OriginalString}' could not be determined";
+                    }
+                }
+            }
+
+            if (stream == null && failureReason == null)
+            {
+                failureReason = $"File '{uri.LocalPath}' could not be opened";
+            }
+
+            if (failureReason != null)
+            {
+                // If we get here, we were unable to determine region character offset, so we have to warn the caller
+                // TODO: add a warning to the list
+            }
+
+            return stream;
+        }
+
+        private StreamReader GetFileStreamReader(Uri uri, out Encoding encoding)
+        {
+            StreamReader reader = null;
+            encoding = null;
+
+            Stream contentStream = GetContentStream(uri, out encoding);
+            if (contentStream != null)
+            {
+                // If we found embedded content, encoding was specified on the file object
+                // If the content is in an extenal file, we let the StreamReader detect the encoding
+                reader = encoding != null ?
```

Comment:
Bug: If you don't have an encoding, there's nothing you can do. As I mentioned in the notes on the pseudo-code I sent: you never guess the encoding from the file stream (it just sniffed the file and made a guess). You can only use the encoding from the SARIF file.
```
private StreamReader GetFileStreamReader(Uri uri, out Encoding encoding)
{
    Stream contentStream = GetContentStream(uri, out encoding);
    return (contentStream != null && encoding != null)
        ? new StreamReader(contentStream, encoding, detectEncodingFromByteOrderMark: false);
        : null;
}
```

Reply:
Forgot about this code. Fixed.